lib/mongo/operation/shared/sessions_supported.rb

```diff
@@ -167,6 +167,17 @@ def add_read_preference(sel, connection)
         # https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-single
         if connection.description.standalone?
           # Read preference is never sent to standalones.
+        elsif connection.description.mongos?
+          # When server is a mongos:
+          # - $readPreference is never sent when mode is 'primary'
+          # - When mode is 'secondaryPreferred' $readPreference is only sent
+          #   when a non-mode field (i.e. tag_sets) is present
+          # - Otherwise $readPreference is sent
+          if read && read.to_doc[:mode] == 'secondaryPreferred'
```

Comment:
Can `if read` condition be pulled up into its own block, and then `read.to_doc` called once to obtain the resulting document (rather than being called many times)?